test/widgets/compose_box_test.dart

```diff
@@ -1407,38 +1412,158 @@ void main() {
       void checkComposeBox({required bool isShown}) => checkComposeBoxIsShown(isShown,
         bannerLabel: zulipLocalizations.errorBannerCannotPostInChannelLabel);
 
-      final narrowTestCases = [
-        ('channel', const ChannelNarrow(1)),
-        ('topic',   eg.topicNarrow(1, 'topic')),
-      ];
-
-      for (final (String narrowType, Narrow narrow) in narrowTestCases) {
-        testWidgets('compose box is shown in $narrowType narrow', (tester) async {
+      const channelNarrow = ChannelNarrow(1);
+      final topicNarrow = eg.topicNarrow(1, 'topic');
+
+      void doTestComposeBoxShown({
+        required Narrow narrow,
+        required bool isChannelSubscribed,
+        required ChannelPostPolicy channelPostPolicy,
+        required UserRole selfUserRole,
+        required bool expected,
+      }) {
+        final description = [
+          narrow.toString(),
+          'channel subscribed? $isChannelSubscribed',
+          'channelPostPolicy: ${channelPostPolicy.name}',
+          'self-user role: ${selfUserRole.name}',
+        ].join(', ');
+        testWidgets(description, (tester) async {
+          final channel = eg.stream(streamId: 1,
+            channelPostPolicy: channelPostPolicy);
           await prepareComposeBox(tester,
             narrow: narrow,
-            selfUser: eg.user(role: UserRole.administrator),
-            streams: [eg.stream(streamId: 1,
-              channelPostPolicy: ChannelPostPolicy.moderators)]);
-          checkComposeBox(isShown: true);
+            selfUser: eg.user(role: selfUserRole),
+            streams: [channel],
+            subscriptions: isChannelSubscribed ? [eg.subscription(channel)] : []);
+          checkComposeBoxIsShown(expected,
+            bannerLabel: isChannelSubscribed
+              ? zulipLocalizations.errorBannerCannotPostInChannelLabel
+              : zulipLocalizations.composeBoxBannerLabelUnsubscribedWhenCannotSend);
         });
+      }
+
+      doTestComposeBoxShown(
```

Comment:
nit: "do" seems redundant; and this calls `testWidgets`, so corresponds to one test case, so it's helpful to use the same "testFoo" naming